test/metabase/lib/filter_test.cljc

```diff
@@ -15,15 +21,79 @@
       (let [f (lib/= venues-category-id-metadata categories-id-metadata)]
         (is (fn? f))
         (is (=? [:=
-                 {:lib/uuid string?}
-                 [:field (meta/id :venues :category-id) {:lib/uuid string?}]
-                 [:field "ID" {:base-type :type/BigInteger, :lib/uuid string?}]]
+                 {:lib/uuid uuid?}
+                 [:field (meta/id :venues :category-id) {:lib/uuid uuid?}]
+                 [:field "ID" {:base-type :type/BigInteger, :lib/uuid uuid?}]]
                 (f {:lib/metadata meta/metadata} -1)))))
     (testing "with query/stage-number, return clause right away"
       (is (=? [:=
-               {:lib/uuid string?}
-               [:field (meta/id :venues :category-id) {:lib/uuid string?}]
-               [:field "ID" {:base-type :type/BigInteger, :lib/uuid string?}]]
+               {:lib/uuid uuid?}
+               [:field (meta/id :venues :category-id) {:lib/uuid uuid?}]
+               [:field "ID" {:base-type :type/BigInteger, :lib/uuid uuid?}]]
+              (lib/= {:lib/metadata meta/metadata}
+                     -1
+                     venues-category-id-metadata
+                     categories-id-metadata))))))
+
+(defn- field-metadata-gen
+  [[_ id-or-name]]
+  (if (integer? id-or-name)
+    (gen/let [[table fields] (gen/elements (for [table (:tables meta/metadata)]
+                                             [(:name table) (map :name (:fields table))]))
+              field (gen/elements fields)]
+      (-> (lib/query meta/metadata table)
+          (lib.metadata/field-metadata table field)))
+    (gen/let [field (gen/elements (map :name (:columns meta/results-metadata)))]
+      (-> (lib/saved-question-query meta/saved-question)
+          (lib.metadata/field-metadata field)))))
+
+(def ^:private filter-expr-gen
+  (gen/let [filter-shape (gen/such-that #(-> % first (= :field) not)
+                                        (mg/generator ::schema.filter/filter))
+            fields (gen/return (set (mbql.u/match filter-shape #{:field :field/unresolved})))
+            field->metadata (gen/return (zipmap fields
+                                                (map #(-> % field-metadata-gen gen/generate)
+                                                     fields)))]
+    (walk/postwalk (fn [form]
+                     (if-let [metadata (field->metadata form)]
+                       metadata
+                       form))
+                   filter-shape)))
```

Comment:
The name is wrong and it's complicated: it generates valid complete filters and then replaces the fields with field metadata so that we can check that these can be added to a query stage.